src/Provider/AbstractProvider.php

```diff
@@ -98,7 +108,15 @@ public function __construct($options = [], HttpAdapterInterface $httpClient = nu
             }
         }
 
-        $this->setHttpClient($httpClient ?: new CurlHttpAdapter());
+        if (empty($collaborators['httpClient'])) {
+            $collaborators['httpClient'] = new CurlHttpAdapter();
+        }
+        $this->setHttpClient($collaborators['httpClient']);
+
+        if (empty($collaborators['randomFactory'])) {
+            $collaborators['randomFactory'] = new RandomFactory();
+        }
+        $this->setRandomFactory($collaborators['randomFactory']);
```

Comment:
Rather than injecting the factory, I think we should allow them to inject the generator itself. So, they use the factory first to get a generator outside of this code, and then they pass that generator into this lib, as long as it follows the proper interface, as suggested by @ircmaxell in https://github.com/thephpleague/oauth2-client/issues/212#issuecomment-97058200